uberfire-security/uberfire-security-management/uberfire-security-management-keycloak/src/main/java/org/uberfire/ext/security/management/keycloak/client/auth/credentials/AuthTokenManager.java

```diff
@@ -0,0 +1,125 @@
+import org.keycloak.OAuth2Constants;
+import org.keycloak.common.util.Time;
+import org.keycloak.util.BasicAuthHelper;
+import org.uberfire.ext.security.management.keycloak.client.auth.TokenManager;
+import org.uberfire.ext.security.management.keycloak.client.auth.TokenService;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+/**
+ * Token manager that uses credentials based authentication settings to manage the access token.
+ * Handles:
+ * - Public / non public clients
+ * - Token refreshments based on OAuth2 token's expiration time.
+ * 
+ * @since 0.9.0
+ */
+public class AuthTokenManager implements TokenManager {
+    
+    private final AuthSettings config;
+    private static final long DEFAULT_MIN_VALIDITY = 30;
+    long expirationTime;
+    long minTokenValidity = DEFAULT_MIN_VALIDITY;
+    AccessTokenResponse accessTokenResponse;
+
+    public AuthTokenManager(AuthSettings config) {
+        this.config = config;
+    }
+    
+    @Override
+    public void grantToken() {
+        MultivaluedMap<String, String> mvm = new CaseInsensitiveMap<String>();
+        mvm.putSingle(OAuth2Constants.GRANT_TYPE, "password");
+        mvm.putSingle("username", config.getUsername());
+        mvm.putSingle("password", config.getPassword());
+
+
+        boolean isPublic = config.isPublicClient();
+
+        String authorization = "";
+        if ( isPublic ) {
+            mvm.putSingle(OAuth2Constants.CLIENT_ID, config.getClientId());
+        }  else {
+            authorization = BasicAuthHelper.createHeader(config.getClientId(), config.getClientSecret());
+        }
+        
+        TokenService client = createTokenService();
+        AccessTokenResponse response = client.grantToken(config.getRealm(), authorization, mvm);
+
+        int requestTime = Time.currentTime();
+        expirationTime = requestTime + response.getExpiresIn();
+        this.accessTokenResponse = response;
+    }
+
+    protected void refreshToken() {
+        MultivaluedMap<String, String> mvm = new CaseInsensitiveMap<String>();
+        mvm.putSingle(OAuth2Constants.GRANT_TYPE, "refresh_token");
+        mvm.putSingle("refresh_token", accessTokenResponse.getRefreshToken());
+
+
+        boolean isPublic = config.isPublicClient();
+
+        String authorization = "";
+        if ( isPublic ) { // if client is public access type
+            mvm.putSingle(OAuth2Constants.CLIENT_ID, config.getClientId());
+        } else {
+            authorization = BasicAuthHelper.createHeader(config.getClientId(), config.getClientSecret());
+        }
+
+        try {
+            
+            TokenService client = createTokenService();
+            AccessTokenResponse response = client.grantToken(config.getRealm(), authorization, mvm);
+
+            int requestTime = Time.currentTime();
+            expirationTime = requestTime + response.getExpiresIn();
+            this.accessTokenResponse = response;
+            
+        } catch (BadRequestException e) {
```

Comment:
@romartin to say somebody else is doing it as well isn't the best argument ;). Like I wrote, you will only get a BadRequestException if the request is syntactically wrong which based on your description isn't what you want. You would have to make sure you get the right status code and if not fall back to ```grantToken```. You can use RestEasy's ClientResponse type for that. It's important to know what the REST endpoint you're invoking is actually returning in case of problems. You won't catch those problems by simply catching BadRequestException as this is only for malformed requests (HTTP 400 indicates a client error).